presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergParquetPrimitiveTypeConverter.java

```diff
@@ -0,0 +1,123 @@
+import static org.apache.parquet.schema.LogicalTypeAnnotation.decimalType;
+import static org.apache.parquet.schema.LogicalTypeAnnotation.intType;
+import static org.apache.parquet.schema.LogicalTypeAnnotation.stringType;
+import static org.apache.parquet.schema.LogicalTypeAnnotation.timeType;
+import static org.apache.parquet.schema.LogicalTypeAnnotation.timestampType;
+import static org.apache.parquet.schema.PrimitiveType.PrimitiveTypeName;
+import static org.apache.parquet.schema.Types.PrimitiveBuilder;
+import static org.apache.parquet.schema.Types.optional;
+
+public class IcebergParquetPrimitiveTypeConverter
+        implements PrimitiveTypeConverter
+{
+    public static final IcebergParquetPrimitiveTypeConverter INSTANCE = new IcebergParquetPrimitiveTypeConverter();
+
+    private IcebergParquetPrimitiveTypeConverter() {}
+
+    @Override
+    public Type getPrimitiveType(io.prestosql.spi.type.Type type, String name)
+    {
+        return primitive(type).named(name);
+    }
+
+    private PrimitiveBuilder<PrimitiveType> primitive(io.prestosql.spi.type.Type type)
+    {
+        if (BOOLEAN.equals(type)) {
+            return optional(PrimitiveTypeName.BOOLEAN);
+        }
+        if (BIGINT.equals(type)) {
+            return optional(PrimitiveTypeName.INT64)
+                    .as(intType(64, true));
+        }
+        if (INTEGER.equals(type)) {
+            return optional(PrimitiveTypeName.INT32)
+                    .as(intType(32, true));
+        }
+        if (SMALLINT.equals(type)) {
+            return optional(PrimitiveTypeName.INT32)
+                    .as(intType(16, true));
+        }
+        if (TINYINT.equals(type)) {
+            return optional(PrimitiveTypeName.INT32)
+                    .as(intType(8, true));
+        }
+        if (REAL.equals(type)) {
+            return optional(PrimitiveTypeName.FLOAT);
+        }
+        if (DOUBLE.equals(type)) {
+            return optional(PrimitiveTypeName.DOUBLE);
+        }
+        if (type instanceof DecimalType) {
+            DecimalType decimalType = (DecimalType) type;
+            if (decimalType.getPrecision() <= 9) {
+                return optional(PrimitiveTypeName.INT32)
+                        .as(decimalType(decimalType.getScale(), decimalType.getPrecision()));
+            }
+            if (decimalType.getPrecision() <= 18) {
+                return optional(PrimitiveTypeName.INT64)
+                        .as(decimalType(decimalType.getScale(), decimalType.getPrecision()));
+            }
+            return optional(PrimitiveTypeName.FIXED_LEN_BYTE_ARRAY)
+                    .length(decimalRequiredBytes(decimalType.getPrecision()))
+                    .as(decimalType(decimalType.getScale(), decimalType.getPrecision()));
+        }
+        if (DATE.equals(type)) {
+            return optional(PrimitiveTypeName.INT32)
+                    .as(dateType());
+        }
+        if (TIMESTAMP.equals(type)) {
```

Comment:
The `TIMESTAMP` constant uses the default precision 3, milliseconds. You probably want to check the precision of the timestamp type. https://github.com/prestosql/presto/blob/master/presto-spi/src/main/java/io/prestosql/spi/type/TimestampType.java#L66

Reply:
This morning @electrum and I decided that introducing the PrimitiveTypeConverter was the wrong approach.  We realized all we really needed to create a ParquetFileWriter was `Map<List<String>, Type>` of the names of primitive type columns and their Presto types, and the MessageType created by `ParquetSchemaUtil.convert()`.

So the code above no longer exists, but thanks the pointer.